### PR TITLE
[Product images] Cache images to reduce network requests

### LINKED ISSUE #11898: [Product image] Cache product images to reduce network load
#### Problem
We load the product images every time we open the product form. There are also multiple requests sent for the same image URL.

I noticed this behaviour while investigating a `WatchdogTermination` Sentry issue. Internal - peaMlT-gA-p2#comment-851

This repeated loading of images might create memory issues when loading a product with many large images. 

#### Solution
- Use cache to store and reuse product images.
- Check and stop sending duplicate network requests for the same image.

#### Screen recording

https://github.com/woocommerce/woocommerce-ios/assets/524475/bf013022-e568-4bae-9cff-45659a7cd1cd

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Closes: #11898

### What? 
We load the product images every time we open the product form. There are also multiple requests sent for the same image URL.

https://github.com/woocommerce/woocommerce-ios/assets/524475/bf013022-e568-4bae-9cff-45659a7cd1cd

I noticed this behaviour while investigating a `WatchdogTermination` Sentry issue. Internal - peaMlT-gA-p2#comment-851

### Why? 

- The repeated loading of images might create memory issues when loading a product with many large images.
- We should stop sending duplicate network requests for the same product image. 

### How? 
- Use `DefaultImageService`'s caching setup to store and reuse product images.
- Rewrite the request image into `async` for better error handling and cancellation. 
- Clear the image cache upon logout.

## Testing instructions

We need to test two things in this PR.
- Product images can be viewed/edited as before.
- Product images should not be fetched from the network each time the product form is opened. (Should be cached instead)

Recommended testing setup
- I recommend setting up [Proxyman](https://proxyman.io/) and running builds from `trunk` and this branch in two simulators. 
- Open a product from the `trunk` build. Check Proxyman to see the image-related network requests.
- Open the same product from this branch's build. Check Proxyman to see the image-related network requests. The images should be cached and reused.
- Follow the same process while testing other product image screens/features. 

Steps
- Create a JN site
- Add a few products with many images
- Login into the app
- Open the products tab and open product and ensure that product images can be viewed as before.
- Try adding/removing product images and validate that it works as before.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/524475/fa9f7052-6ec6-438c-8401-697b0dec505d

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Media/DefaultProductUIImageLoader.swift [hunk 1]
```diff
@@ -51,31 +54,41 @@ final class DefaultProductUIImageLoader: ProductUIImageLoader {
         }
     }
 
-    deinit {
-        activeImageTasks.forEach { $0.cancel() }
-        activeImageTasks.removeAll()
-    }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Products/Media/DefaultProductUIImageLoader.swift:57]: ❓ Just wondering: don't we need to cancel all background tasks when the loader is deallocated anymore?

**Review by itsmeichigo — APPROVED** [review]
This works as expected, the images in product forms load instantly after the first try :shipit:

File: WooCommerce/Classes/ViewRelated/Products/Media/DefaultProductUIImageLoader.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Products/Media/DefaultProductUIImageLoader.swift:57]: Good question. 

We avoid taking ownership of the `Task` here as `ProductImageCollectionViewCell` owns the `Task`, and the task will be deallocated/cancelled when the collection view is deallocated. https://github.com/woocommerce/woocommerce-ios/pull/11902/files#diff-435040a6ba1418be8ae0640c6306cb49930877138097038b7c2a06bd361ecd89

Does it sound good to you?

**Comment by selanthiraiyan** [comment]
Thanks for the review and testing, Huong! 🙇 I am enabling auto merge now. Please feel free to add comments/questions, if any.

**Pull request merged**